lib/filterx/expr-regexp.c

```diff
@@ -417,3 +438,290 @@ filterx_expr_regexp_search_generator_new(FilterXExpr *lhs, const gchar *pattern)
+}
+
+static FilterXObject *
+_replace_matches(const FilterXFuncRegexpSubst *self, FilterXReMatchState *state)
+{
+  GString *new_value = scratch_buffers_alloc();
+  PCRE2_SIZE *ovector = NULL;
+  gint pos = 0;
+
+  do
+    {
+      ovector = pcre2_get_ovector_pointer(state->match_data);
+
+      g_string_append_len(new_value, state->lhs_str + pos, _start_offset(ovector) - pos);
+      g_string_append(new_value, self->replacement);
+
+      if (_is_zero_length_match(ovector))
+        {
+          g_string_append_len(new_value, state->lhs_str + pos, 1);
+          pos++;
+        }
+      else
+        pos = _end_offset(ovector);
+
+      if (!_match_inner(state, self->pattern, pos))
+        break;
+    }
+  while ((pos < state->lhs_str_len) && self->opts.global);
+
+  // add the rest of the string
+  g_string_append_len(new_value, state->lhs_str + pos, state->lhs_str_len - pos);
+
+  // handle the very last of zero lenght matches
+  if (_is_zero_length_match(ovector))
+    g_string_append(new_value, self->replacement);
+
+  return filterx_string_new(new_value->str, new_value->len);
+}
+
+static FilterXObject *
+_subst_eval(FilterXExpr *s)
+{
+  FilterXFuncRegexpSubst *self = (FilterXFuncRegexpSubst *) s;
+
+  FilterXObject *result = NULL;
+  FilterXReMatchState state;
+  _state_init(&state);
+
+  gboolean matched = _match(self->string_expr, self->pattern, &state);
+  if (!matched)
+    {
+      result = filterx_object_ref(state.lhs_obj);
+      goto exit;
+    }
+
+  if (!state.match_data)
+    {
+      /* Error happened during matching. */
+      result = NULL;
+      goto exit;
+    }
+
+  result = _replace_matches(self, &state);
+
+exit:
+  _state_cleanup(&state);
+  return result;
+}
+
+static FilterXExpr *
+_extract_subst_string_expr_arg(FilterXFunctionArgs *args, GError **error)
+{
+  return filterx_function_args_get_expr(args, 0);
+}
+
+static pcre2_code_8 *
+_extract_subst_pattern_arg(FilterXFuncRegexpSubst *self, FilterXFunctionArgs *args, GError **error)
+{
+  const gchar *pattern = filterx_function_args_get_literal_string(args, 1, NULL);
+  if (!pattern)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "argument must be a string literal: pattern. " FILTERX_FUNC_REGEXP_SUBST_USAGE);
+      return NULL;
+    }
+
+  return _compile_pattern(pattern, !self->opts.nojit, self->opts.match_opts);
+}
+
+static gchar *
+_extract_subst_replacement_arg(FilterXFunctionArgs *args, GError **error)
+{
+  const gchar *replacement = filterx_function_args_get_literal_string(args, 2, NULL);
+  if (!replacement)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "argument must be a string literal: replacement. " FILTERX_FUNC_REGEXP_SUBST_USAGE);
+      return NULL;
+    }
+
+  return g_strdup(replacement);
+}
+
+static gboolean
+_extract_optional_flags(FilterXFuncRegexpSubst *self, FilterXFunctionArgs *args, GError **error)
```

Comment:
I like the idea of separating the matching options to a separate entity, we should lean into it more IMO.
`FilterXRegexpFlags` could be a separate class, and this function should operate on it only. This way we will be able to reuse it for other functions as well, e.g. for `regexp_search()` when it becomes a real function.